keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+- Keep the core DRA functionality from [KEP
+  #3063](https://github.com/kubernetes/enhancements/issues/3063) working as
+  defined there for Kubernetes 1.29. Cluster autoscaling is not always needed
+  and in-tree numeric model(s) are not expected to cover all use case, so the
+  flexibility offered by opaque parameters remains relevant.
+
+### Non-Goals
+
+- Support semantic class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Define exactly one semantic model that covers all current and future resources.
+  There cannot be a "one size fits all" solution.
+
+## Proposal
+
+Each node publishes information about the nearly-opaque resources it currently
+has. Each driver manages one or more resource instance per node. An instance
+could be a single discreet piece of hardware like a PCI card or a pool of such
+cards. Each instance has certain attributes. These can be qualitative (labels
+or fields for comparison) and quantitative (capacity that gets reduced when
+allocated for a claim). These attributes are described by the "attributes type"
+of the semantic models. Each instance can combine attributes from different
+semantic models. This way, each model can focus on one aspect, like counting
+items or selection via labels.
+
+Workloads make claims to request resources with claim parameters that use CRDs
+provided together with a DRA driver. As soon as such a custom CR is created, a
+CRD controller of the DRA driver must convert it into a in-tree claim parameter
+object that uses the "parameter types" defined by the semantic models.
+
+The scheduler receives both the node resource information and the claim
+parameters. It uses the in-tree claim parameters to identify and allocate
+suitable resources. It captures all information that the DRA driver kubelet
+plugin will need in the "allocation types" of the semantic models and stores
+that in the claim status. Then it schedules the pod.
+
+The DRA driver kubelet plugin receives that information when the pod is about to
+start. It checks that the hardware is still available and healthy, then
+configures it for the pod as specified in the allocation types.
+
+### Publishing node resources
+
+A semantic model defines an attributes type. This is a Go type that gets
+embedded inside the resource.k8s.io API, not a separate API kind. The kubelet
+queries the registered DRA driver kubelet plugins for resource information
+through a new gRPC call. A streaming gRPC API allows drivers to update the
+capacity information.
+
+For example, a simple counter could be described in a `counter` Go package as:
+```
+type Attributes struct {
+    // Some meta information just for logging and debugging. Not used for counting.
+	metav1.ObjectMeta
+
+	Count int64
```

Comment:
Maybe it's a problem with the terminology (sorry, yes, again), but I fail to see how a `Count` of something qualifies as an attribute of it (unless this is somehow only used for selection in some way, and *not* for resource accounting).

Reply:
To be clear -- at least in my mind -- the sole purpose of attributes should be so that one can do selection against a given instance type later on. Something else needs to exist outside of this to declare how many instances that have these attributes are available on a given node.

Reply:
Agree - I tried to avoid "selection" because of the baggage of k8s selector semantics, but I think we can streamline it as you suggest.

"qualitative" "constraints" based on descriptive "attributes" get "resolved" or "filtered" or "evaluated" against potential instances.

"quantitative" "requests" for "capacity" can "consume" the published "capacity".